source-firestore/discovery.go

```diff
@@ -0,0 +1,237 @@
+				docsCh: docsCh,
+				count:  0,
+				open:   true,
+			}
+
+			eg.Go(func() error {
+				var collection = request.collection
+				var err error
+				var schema schema_inference.Schema
+				log.WithField("collection", collection).Debug("schema_inference.Run")
+				schema, err = schema_inference.Run(ctx, logEntry, docsCh)
+				if err != nil {
+					return fmt.Errorf("schema inference: %w", err)
+				}
+
+				resourceJSON, err := json.Marshal(resource{Path: collection})
+				if err != nil {
+					return fmt.Errorf("serializing resource json: %w", err)
+				}
+
+				log.WithField("collection", collection).Debug("finished schema inference")
+				response.Bindings = append(response.Bindings, &pc.DiscoverResponse_Binding{
+					RecommendedName:    pf.Collection(collectionRecommendedName(collection)),
+					ResourceSpecJson:   resourceJSON,
+					DocumentSchemaJson: schema,
+					KeyPtrs:            []string{"/" + firestore.DocumentID},
+				})
+				return nil
+			})
+
+			docsCh <- request.document
+		} else {
+			if process.open {
+				log.WithField("collection", request.collection).WithField("count", process.count).Debug("channel exists, sending document")
+				process.docsCh <- request.document
+				process.count = process.count + 1
+				if process.count > DISCOVER_DOC_LIMIT {
+					process.open = false
+					close(process.docsCh)
+				}
+			} else {
+				log.WithField("collection", request.collection).Debug("channel is closed")
+			}
+		}
+	}
+
+	return nil
+}
+
+// Go through all collections and discover them recursively. Spawn a discoverRoutine for handling management
+// of inference channels
+func discoverCollections(ctx context.Context, client *firestore.Client, response *pc.DiscoverResponse) error {
+	var collections = client.Collections(ctx)
+	var ch = make(chan inferenceRequest)
+	var firestoreGroup = new(errgroup.Group)
+	var routineGroup = new(errgroup.Group)
+
+	var inferenceChannels = map[string]*inferenceProcess{}
+
+	for {
+		var collection, err = collections.Next()
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		firestoreGroup.Go(func() error {
+			return discoverCollection(ctx, ch, inferenceChannels, response, collection)
+		})
+	}
+
+	routineGroup.Go(func() error { return discoveryRoutine(ctx, response, routineGroup, ch, inferenceChannels) })
+
+	if err := firestoreGroup.Wait(); err != nil {
+		return err
+	}
+
+	close(ch)
+	for _, process := range inferenceChannels {
+		if process.open {
+			close(process.docsCh)
+		}
+	}
+
+	if err := routineGroup.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Recursively discover collections
+func discoverCollection(
+	ctx context.Context,
+	ch chan inferenceRequest,
+	inferenceChannels map[string]*inferenceProcess,
+	response *pc.DiscoverResponse,
+	collection *firestore.CollectionRef,
+) error {
+	var process, exists = inferenceChannels[collectionToResourcePath(collection.Path)]
```

Comment:
there's a data race over `inferenceChannels`, since it can be concurrently mutated by `discoveryRoutine`.

There's also a race over `process.open` which is read here but concurrently modified elsewhere.

Can you refactor this to accept a context, which is cancelled by discoverRoutine when it's read enough from this collection?

Reply:
Refactored, let me know if it makes sense now

Reply:
You still have a race over `inferenceChannels`, since discoveryRoutine adds new entries to it while this concurrently access it. Try building / running with --race detection?

Reply:
I've just pushed a new commit to this PR which reimplements the discovery process in a way that seemed cleaner to me (a single goroutine is spawned to handle each _resource_ which is discovered, and is fed collections to process over a channel from the parent resource's goroutine which acts as a work queue).

There's still a map that gets accessed concurrently from various workers though, so I've added a mutex to guard that.